core/store/src/parallel_iter.rs

```diff
@@ -0,0 +1,692 @@
+            WorkItem::Lookup(_, range) => {
+                write!(f, "keys {}..{}", range.start, range.end)
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+struct IterationRange {
+    start: Vec<u8>,
+    end: Vec<u8>, // empty means unbounded
+}
+
+impl IterationRange {
+    fn new(start: Vec<u8>, end: Vec<u8>) -> Self {
+        assert!(start < end || end.is_empty());
+        Self { start, end }
+    }
+
+    /// Returns the smallest possible key that is greater than the given key.
+    fn next_key(key: &[u8]) -> Vec<u8> {
+        let mut key = key.to_vec();
+        key.push(0);
+        key
+    }
+
+    /// Divides the range into smaller ranges. The implementation can be anything as long as
+    /// this will exponentially divide the key space.
+    fn divide(&self) -> Vec<Self> {
+        // First find the common prefix; the resulting ranges will always start with these,
+        // so just ignore these for the remaining processing.
+        let mut common_prefix = Vec::new();
+        for (a, b) in self.start.iter().zip(self.end.iter()) {
+            if a == b {
+                common_prefix.push(*a);
+            } else {
+                break;
+            }
+        }
+
+        // Remove the common prefix.
+        let start = self.start[common_prefix.len()..].to_vec();
+        let end = self.end[common_prefix.len()..].to_vec();
+
+        // We'll find some split points and use these to split the ranges.
+        let mut split_points = Vec::new();
+
+        // If the start is the beginning, divide based on the first character.
+        if start.is_empty() {
+            if end.len() == 1 {
+                // If the end is just one byte, then that byte is exclusive. There exists a corner
+                // case where the range is "..00" (i.e. the end is a single zero byte); this range
+                // is unsplittable as it only contains one element. But it's OK because we'll end
+                // up with no splitting point and we'll return the single range.
+                for byte in 0..end[0] {
+                    let point = vec![byte];
+                    split_points.push(point);
+                }
+            } else {
+                // Otherwise the end byte is inclusive.
+                let first_character_max = end.get(0).copied().unwrap_or(0xff);
+                for byte in 0..=first_character_max {
+                    let point = vec![byte];
+                    split_points.push(point);
+                }
+            }
+        }
+        // If there's no end, then divide based on the first byte of start that is not 0xff.
+        else if end.is_empty() {
+            let mut byte_not_ff = 0;
+            for (i, byte) in start.iter().enumerate() {
+                if *byte != 0xff {
+                    byte_not_ff = i;
+                    break;
+                }
+            }
+            for byte in start[byte_not_ff] + 1..=0xff {
```

Comment:
what if end is empty and start == [0xff] ? then forgetting about what we want to do about splitting, this `start[byte_not_ff] + 1` will panic with overflow

this triggers the panic:

```diff
diff --git a/core/store/src/parallel_iter.rs b/core/store/src/parallel_iter.rs
index eb7c6d7b9..ce3998c2c 100644
--- a/core/store/src/parallel_iter.rs
+++ b/core/store/src/parallel_iter.rs
@@ -489,6 +489,9 @@ mod tests {
 
     #[test]
     fn test_iteration_range_divide() {
+        let ranges = IterationRange::from_str("ff..").unwrap().divide();
+        println!("{}", ranges[0].to_string());
+
         let ranges = IterationRange::from_str("..").unwrap().divide();
         assert_eq!(ranges.len(), 257);
         assert_eq!(ranges[0].to_string(), "..00");
```

Reply:
Haha... nice catch. Well I'm removing all this code so, phew. This code is so messy. I wonder if there's a good way to make it cleaner.